govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/coherence/metric.py

```diff
@@ -0,0 +1,193 @@
+        ):
+            self.error = None
+            prompt = self.evaluation_template.evaluate(
+                user_input=test_case.input or "",
+                actual_output=test_case.actual_output or "",
+                evaluation_steps=self.evaluation_steps,
+                rubric=self.rubric,
+                score_range=self.score_range,
+                additional_context=self.additional_context,
+            )
+
+            try:
+                judgement = await self._generate_result_from_model(
+                    prompt, schema=CoherenceJudgement
+                )
+            except Exception as exc:
+                self.error = f"Failed to generate coherence judgement: {exc}"
+                self.score = float("nan")
+                self.success = False
+                self.reason = None
+                logging.error(self.error, exc_info=exc)
+                return float("nan")
+
+            raw_score = self._coerce_and_clamp_score(judgement.score)
+            self.raw_score = raw_score
+            self.score = self._normalize(raw_score, self.score_range)
+            cleaned_reason = judgement.reason.strip()
+            self.reason = cleaned_reason if self.include_reason else None
+            self.success = self.is_successful()
+
+            capture_metric_type(
+                self.__name__, async_mode=True, in_component=_in_component
+            )
+
+            verbose_reason = cleaned_reason if self.include_reason else "Reason omitted"
+            self.verbose_logs = construct_verbose_logs(
+                self,
+                steps=[
+                    f"Raw Score: {raw_score}",
+                    f"Score: {self.score}",
+                    f"Reason: {verbose_reason}",
+                ],
+            )
+
+            return self.score
+
+    async def _generate_result_from_model(
+        self, prompt: str, schema: Type[CoherenceJudgement]
+    ) -> CoherenceJudgement:
+        if self.using_native_model:
+            raw_result, raw_cost = await self.model.a_generate(prompt, schema=schema)
+            if not isinstance(raw_result, schema):
+                raise TypeError("Generated result does not match the expected schema.")
+            self._increment_cost(
+                raw_cost if isinstance(raw_cost, (int, float)) else None
+            )
+            return raw_result
+
+        try:
+            result = await self.model.a_generate(prompt, schema=schema)
+            if isinstance(result, schema):
+                return result
+            raise TypeError("Generated result does not match the expected schema.")
+        except TypeError:
+            raw_response = await self.model.a_generate(prompt)
+            data = trimAndLoadJson(raw_response, self)
+            return schema(**data)
+
+    def _normalize(self, raw_score: int, score_range: tuple[int, int]) -> float:
+        lo, hi = score_range
+        if hi <= lo:
+            return 0.0
```

Comment:
I don't think we want to have conditional logic on score_range since that's just a constant we already control